sql/vector_mhnsw.cc

```diff
@@ -237,28 +237,49 @@ struct FVector
   static float dot_product(const int16_t *v1, const int16_t *v2, size_t len)
   {
     // Using vector long long for int64_t accumulation
-    vector long long ll_sum= {0, 0};
-    // Round up to process full vector, including padding
-    size_t base= ((len + POWER_dims - 1) / POWER_dims) * POWER_dims;
+    vector long long ll_sum1= {0, 0};
+    vector long long ll_sum2= {0, 0};
+    vector long long ll_sum3= {0, 0};
+    vector long long ll_sum4= {0, 0};
 
-    for (size_t i= 0; i < base; i+= POWER_dims)
-    {
-      vector short x= vec_ld(0, &v1[i]);
-      vector short y= vec_ld(0, &v2[i]);
-
-      // Vectorized multiplication using vec_mule() and vec_mulo()
-      vector int product_hi= vec_mule(x, y);
-      vector int product_lo= vec_mulo(x, y);
+    size_t elements_per_loop= POWER_dims * 4;
 
-      // Extend vector int to vector long long for accumulation
-      vector long long llhi1= vec_unpackh(product_hi);
-      vector long long llhi2= vec_unpackl(product_hi);
-      vector long long lllo1= vec_unpackh(product_lo);
-      vector long long lllo2= vec_unpackl(product_lo);
+    // Round up to nearest multiple of elements_per_loop (with padding)
+    size_t base= ((len + elements_per_loop - 1) / elements_per_loop) *
+                 elements_per_loop;
```

Comment:
You overrun buffers here, allocation is `POWER_dims` aligned, not `POWER_dims * 4`. Feels like you have to raise `POWER_bytes= 128 / 8 * 4`.